### PR TITLE
Port/1.12 hotifxes

### PR DESCRIPTION
**Pull request by Raycoms**
Several important bugfixes with furnaces and job assignment

Review please

### CODE REVIEW

File: src/main/java/com/minecolonies/coremod/colony/CitizenData.java [hunk 1]
```diff
@@ -233,6 +233,70 @@ public void markDirty()
         colony.markCitizensDirty();
     }
 
+    @Override
+    public boolean equals(final Object o)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/colony/CitizenData.java:237]: MAJOR Reduce the number of returns of this method 11, down to the maximum allowed 6. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1142)

File: src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java [hunk 2]
```diff
@@ -324,7 +323,8 @@ public void onWorldTick(@NotNull final TickEvent.WorldTickEvent event)
 
         // If we have no active worker, grab one from the Colony
         // TODO Maybe the Colony should assign jobs out, instead?
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java:325]: INFO Complete the task associated to this TODO comment. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1135)

File: src/main/java/com/minecolonies/coremod/colony/buildings/BuildingHome.java [hunk 3]
```diff
@@ -164,6 +164,7 @@ public void removeCitizen(@NotNull final CitizenData citizen)
         {
             citizen.setHomeBuilding(null);
             residents.remove(citizen);
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/colony/buildings/BuildingHome.java:166]: MINOR This call to "remove()" may be a performance hot spot if the collection is large. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2250)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 78 issues
* ![MAJOR][MAJOR] 34 major
* ![MINOR][MINOR] 41 minor
* ![INFO][INFO] 3 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [CitizenData.java#L26](https://github.com/Minecolonies/minecolonies/blob/024b00af824763956fd87ce864d0daa8435383a4/src/main/java/com/minecolonies/coremod/colony/CitizenData.java#L26): class "CitizenData" has 45 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [CitizenData.java#L518](https://github.com/Minecolonies/minecolonies/blob/024b00af824763956fd87ce864d0daa8435383a4/src/main/java/com/minecolonies/coremod/colony/CitizenData.java#L518): Use the "equals" method if value comparison was intended. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1698)
1. ![MAJOR][MAJOR] [CitizenData.java#L523](https://github.com/Minecolonies/minecolonies/blob/024b00af824763956fd87ce864d0daa8435383a4/src/main/java/com/minecolonies/coremod/colony/CitizenData.java#L523): Use the "equals" method if value comparison was intended. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1698)
1. ![MAJOR][MAJOR] [CitizenData.java#L578](https://github.com/Minecolonies/minecolonies/blob/024b00af824763956fd87ce864d0daa8435383a4/src/main/java/com/minecolonies/coremod/colony/CitizenData.java#L578): Use the "equals" method if value comparison was intended. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1698)
1. ![MAJOR][MAJOR] [CitizenData.java#L582](https://github.com/Minecolonies/minecolonies/blob/024b00af824763956fd87ce864d0daa8435383a4/src/main/java/com/minecolonies/coremod/colony/CitizenData.java#L582): Use the "equals" method if value comparison was intended. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1698)
1. ![MAJOR][MAJOR] [AbstractBuildingWorker.java#L144](https://github.com/Minecolonies/minecolonies/blob/024b00af824763956fd87ce864d0daa8435383a4/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java#L144): Remove this unused method parameter "stack". [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1172)
1. ![MAJOR][MAJOR] [AbstractBuildingWorker.java#L188](https://github.com/Minecolonies/minecolonies/blob/024b00af824763956fd87ce864d0daa8435383a4/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java#L188): Either log or rethrow this exception. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1166)
1. ![MAJOR][MAJOR] [AbstractBuildingWorker.java#L352](https://github.com/Minecolonies/minecolonies/blob/024b00af824763956fd87ce864d0daa8435383a4/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java#L352): Reduce this class from 52 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [BuildingHome.java#L331](https://github.com/Minecolonies/minecolonies/blob/024b00af824763956fd87ce864d0daa8435383a4/src/main/java/com/minecolonies/coremod/colony/buildings/BuildingHome.java#L331): Reduce this class from 38 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [EntityCitizen.java](https://github.com/Minecolonies/minecolonies/blob/024b00af824763956fd87ce864d0daa8435383a4/src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java): This file has 1.388 lines, which is greater than 1.000 authorized. Split it into smaller files. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS00104)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Review by marchermans — APPROVED** [review]

**Pull request merged**